### DISCUSSION TITLE
Connection terminated unexpectedly

### DISCUSSION THREAD

**Discussion by galaczi** [adonisjs/core#3887] [Help] [answered]
Before I open an issue in the Lucid repo, I would like to ask if anyone has encountered random ``Connection terminated unexpectedly`` errors. It happens on a single route, so it's probably the query that does it. It seems to happen on the first visit on that route,  repeated visits are fine.

```
"msg": "select \"locations\".*, \"profile_at_student_locations\".\"profile_id\" 
as \"pivot_profile_id\", 
\"profile_at_student_locations\".\"location_id\" 
as \"pivot_location_id\" 
from \"locations\" 
inner join \"profile_at_student_locations\"
 on \"locations\".\"id\" = \"profile_at_student_locations\".\"location_id\" 
where \"profile_at_student_locations\".\"profile_id\" in ($1) 
- Connection terminated unexpectedly"
```

Edit: it's happening with other routes as well. Looks like after some idle time the connection drops. The first visit reestablishes the connection and then everything is fine. Where can I tweak the pool settings? Should I?

**Comment by thetutlage** [comment, marked as answer]
There are many reasons for a connection to terminate and impossible for us to know why. So, you basically have to find out the root cause for this and then report an issue.

**Reply by galaczi** [reply]
After some more research, my issue seems to be that the Docker Swarm overlay network closes idle tcp connections after 900 seconds. I should set up a keepalive interval lower than that. I poked around in ``database.ts`` but couldn't find the right place to do it. Is it possible? Otherwise I should set it on the host OS.

**Reply by thetutlage** [reply]
Yeah, I think the database should not drop connections abruptly. One thing you can try it set the `pool.min = 0`. But that way you will be paying a penalty of making a database connection every time you run a query.

It is better to have some connections warmed up all the time. Maybe fix Docker swarm to keep some connections alive

**Reply by galaczi** [reply]
Adding this to the postgres service inside docker-compose.yml solved the issue completely:

```
    command: postgres -c tcp_keepalives_idle=60 -c tcp_keepalives_interval=60 -c tcp_keepalives_count=60

```
